### ISSUE TITLE
Implement other GUI extensions

### ISSUE THREAD

**Issue by smolck** [neovide/neovide#768]
Currently Neovide implements `ext_multigrid`, but that's not the only GUI extension of Neovim; there's also `ext_cmdline`, `ext_popumenu`, `ext_tabline`, and others, which to my knowledge are not implemented. Implementing these would allow for more control over the display of the cmdline, pmenu, tabline, etc., and also could improve the look of things.

For instance, with `ext_cmdline`, the cmdline could be displayed more in the center of the screen, as it is in GNvim, GoNeovim, and others, which is arguably a nicer experience (don't have to look to the corner of the screen when opening/using the cmdline).

Considering the README says that Neovide should "functionally . . . act like the terminal UI," I'm not certain on the sentiment about these sorts of features, and so totally understand if this is not wanted, but figured I might as well open an issue on the topic.

**Comment by shaunsingh** [comment]
This is actually something we're interesting in working on soon. I'll close this issue in favor of these 

https://github.com/Kethku/neovide/issues/233 (tabs and statusline), https://github.com/Kethku/neovide/issues/324 (command palette) and https://github.com/Kethku/neovide/issues/669 (minimap_

**Comment by Kethku** [comment]
I'm actually going to reopen this to encourage more conversation. @smolck something I've discussed here: https://github.com/Kethku/neovide/issues/669 is the idea of exposing graphics APIs to the vim side. Ideally these would be very simple vector graphics style drawing commands which take a window and potentially have a way to look up a grid location in pixel coordinates. Then implementing some forms of ui affordances is up to plugin authors.

To take this idea even further, if we enable font sizing on a window by window basis (not there yet, but within striking distance) we could even expose auto commands for drawing the command pallette/messages/popup menu. That way we'd get some of the benefit of those types of custom controls while still keeping things very vim centric. I don't have my heart set on this idea, but its got some appealing aspects because this sort of pattern could encourage other neovim guis to expose the same sort of thing and give more flexibility to plugin authors for crafting awesome experiences.

You mention the "functionally act like the terminal UI" comment in the readme. This was mostly a response to an issue I recognized with a couple of the existing neovim guis where it doesn't "feel" like neovim. Often they had ui components that I'm sure some people liked, but to me felt like add ons with minor improvements in terms of usability. Neovide isn't perfect in this regard (the cursor trails are purely silly) but when I consider new features to add to it, I'm looking for big usability gains. Some examples:
- the cursor animation allows users (and observers of people using the editor) to more easily track where the cursor has gone
- the window animations do the same thing when windows move around. Its much easier to understand how windows have changed if they subtly animate into position
- ditto scroll animations. Way easier to cycle through search results when the animation tells you vaguely where in the document you jumped to
-  and lastly the blurred window background increases separation between the floating window and the main grid while still letting the upstream transparent window feature shine

My point is that these improvements are pretty targeted at an existing pain point. Adding infrastructure and maintenance burden ideally should come with a feature or friction improvement. Beyond that, I'm all ears!

**Comment by clason** [comment]
This is a bit of a "tag team" comment, so take it with a grain of salt, but:

1. I agree that exposing more configurability to neovim is good, but I'd do it in terms of Lua functions. (Not for the hype, but because Lua is more composable, and that's where neovim itself is moving for its newer API extensions.)
2. Continuing from that, I would focus less on low-level drawing primitives and more on high-level UI components which users can "opt-in" to slot in for native functions. As a primitive (hah!) example, neovim recently introduced a `vim.notify()` function that is slowly being used instead of pushing stuff to `:messages`. (Internally, it still does just that, but that makes it easy at some point for a user to just write `lua vim.notify = neovide.notify` to replace this globally with a native GUI notification center or whatnot). Longer-term, there's https://github.com/mjlbach/neovim-ui, which is pretty threadbare for now but will be a focus during the next development cycle or so. (So input from GUI devs is very useful now ;))

**Comment by Kethku** [comment]
@clason I've recently moved my vim config over to lua, and I agree it does appear to be strictly better and I really like that approach of using a lua function which can be dynamically overwritten. I hadn't thought about that, but that seems like a pretty slick way to go about it.

As for low-level drawing primitives vs high level components, I think my answer is to do both. Things like minimaps are really hard to implement without some form of low level drawing. Similarly, its really hard to render images in most neovim guis. So I would like to use both of those things as motivating use cases.

However I also find neovim-ui to be VERY interesting. I'm not sure how best I can help out with that effort as I don't actually write much in the way of neovim plugins (yet). I'll do some looking at the neovim-ui project in the coming days and get back to you

**Comment by clason** [comment]
Yes, one doesn't preclude the other. You can have a lower-level drawing API for plugins _and_ provide basic components built from them that a future neovim specification describes (`notify`, `picker`, etc. -- this is being discussed).

And I think the input from the "business end" side of things (GUIs that would implement their own version of these components, all shiny and chrome) would be just as valuable as input from "consumers" like plugins to map out the space of possibilities.

**Comment by kiuKisas** [comment]
Let's summarize a bit and bring back the discussion, as I feel like we have two differents things here, but I might be wrong. Let me explain as I understand it:
## Externalize UI
As of today, according to [nvim doc](https://neovim.io/doc/user/ui.html), we have these external uis:
### Done
- [x] `ext_multigrid`, allow to manage per-window externally. we already got this one :)
- [x] `ext_linegrid`,seems to be implemented and required for multigrid ?
### Priority ?
- [ ] `ext_cmdline`: externalize the cmdline.
 Can be cool, and probably the priority between all of external uis not implemented yet. [doc](https://neovim.io/doc/user/ui.html#ui-cmdline). It will ask to impl
 - [ ] `ext_popup` (popupmenu-completion + wildmenu):  we already have some transparency/animation will keeping the native one through `multigrid` support if I understand correctly.  Can be cool too, especially if we externalize the cmdline, being able to put the wildmenu next to it seem to be a must have for exemple.
 ### Later ?
 - [ ] `ext_hlstate`: not sure what we can do with that though, more visual effect on highlight ? Probably not a priority
 - [ ] `ext_messages`: we already have some nvim native solution for good looking notification and desktop native notification. So, not a priority either, but can be nice one day.
 - [ ] `ext_tabline`: same, we already have good looking neovim stuff, let's not focus on it for now.
 
So, according to me, we should focus on `ext_cmdline` and `ext_popup`. It doesn't seems to need the low-level API @Kethku describe. 

Roughly and as far as I can tell, I would say that:
#### Minimal requirement for both
- [ ] An `ext screen` that is not neovim. It should be display on top of `neovim screen` when needed, full transparency by default (but some blur can be cool though).
- [ ] This `ext screen` should contains severals `ext view` (cmdline, popup, maybe something else in the future), and display them according to their position/dimension. 
- [ ] A way to show many  `ext view/ui` on top of neovim screen
- [ ] let user define some basic ui settings for each view (dimension, position.. animation ?)

#### Require for `cmdline`
- [ ] a `prompt widget`, one line interactive text. that can capture input, send it to the neovim api, and updated itself in async way. We probably have someting similar for displaying neovim window ?
- [ ] some implementation that link the `prompt widget` with the cmdline api [(seems doable)](https://neovim.io/doc/user/ui.html#ui-cmdline).

### Require for `popup`
- [ ] a `rofi-like widget` that can interact
- [ ] some implementation between this widget and neovim ui event

#### Enhancement/Possibility
- [ ] more options for `ext view`: font, font-size, so on..
- [ ] let the user create an `external view` based on the widget ?

#### To summarize, we will have 3 main stuffs:
- `2 GUI widgets`: a prompt and rofi-like menu (low)
- `2 external ui/view`: cmdline and popup. (middle)
- an `ext screen` that show external ui on top of neovim ui. (high)

 ## "Neovide UI API"
 So, we will have two things here. Let's call them: `neovide widgets` and `neovide vector API`.
 We don't need the `neovide vector API` yet to have some basic external ui that we can take advantage of (e.g: a center cmdline prompt). In the hierarchie I describe, we can imagine that the `neovide vector API` can become a widget, so we can implement it later.
 This way we can separate 
 
 What do you all thing about it ?

**Comment by ryo33** [comment]
I want to advance the discussion about low-level graphic API.
My initial motivation for this was expressed in a discussion [here](https://discordapp.com/channels/896132176091955261/896264546686222407/1078561544507498506). I would greatly appreciate your thoughts, critiques, and suggestions on the proposed Neovim No-Nonsense Protocol (NNNP).

## *"Neovim No-Nonsence Protocol"* (NNNP)

I'd like to propose "Neovim No-Nonsence Protocol" as the working name for this project to help our discussion. It can also be used as the official name when releasing it if no nice names are found.

## Potential Use Cases

- showing a minimap.
- showing a color picker.
- showing icons without Nerd Fonts.
- showing [pets](https://github.com/giusgad/pets.nvim).

## Not specific to Neovide

The intention to develop NNNP is for providing a graphic protocol across various Neovim plugins and GUI applications. Consequently, we cannot assume how the GUI renders things (Qt, winit, Electron, and more). While the specification does not include "Neovide", we can push forward this domain as the Neovide community. If the protocol is friendly and useful, GUIs and plugins would adopt it.

## But specific to Neovim (widgets are just Vim windows)

If not, we need to reinvent Qt or something, and it would complicated protocol that no one wants to use. To avoid it, we can use windows and buffers as the primitive of NNNP.

Windows and buffers are used to:

- indicate the position, width, height, and z-index of the widget.
- draw the background color.
- represent the lifetime of the widget.
- tell UI what should be rendered using a buffer with a special filetype like `nnnpwidget`.
- receive keyboard events.

## Its architecture

- UI sets a global variable to tell plugins NNNP is available.
- A plugin creates a floating window with a buffer with the special file type.
- The plugin can ask for the dimension (logical size and scale factor) of one character in the grid.
- The plugin provides image data to UI via the buffer.
- UI renders the image.

## Protocol details

- NNNP should support only bitmap format for image data to ensure simplicity and ease of implementation across GUI applications.
- It should support transparency.
- It must support that multiple plugins can use NNNP simultaneously.
- [ ] How to achieve "The plugin can ask for the dimension (logical size and scale factor) of one character in the grid."?

**Comment by Kethku** [comment]
@ryo33 this seems like a pretty cool proposal. If youre still interested in this, could you create a new issue with this stuff captured? I think it got lost in this pretty old issue and so was missed. Closing this one as we have implemented many of the above extensions and intentionally decided not to implement the others.

**Comment by ryo33** [comment]
@Kethku Thanks for your suggestion. I've opened the above issue.
